proposals/0000-type-level-literals.rst

```diff
@@ -0,0 +1,92 @@
+Type-level literals as a separate language extension
+====================================================
+
+.. author:: Ross Paterson
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/536>`_.
+.. sectnum::
+.. contents::
+
+Introduce a language extension ``TypeLevelLiterals`` (implied by ``DataKinds``)
+to turn on the features in the *Type-Level Literals* section of the
+*User's Guide*.
+
+.. _`#102`: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0106-type-data.rst
+
+Motivation
+----------
+Currently type-level literals are turned on by the language extension
+``DataKinds``, but some programmers want to use them without turning on
+datatype promotion.
+
+Indeed, `#102`_ (Define Kinds Without Promotion) will allow programmers to avoid
+``DataKinds`` by using ``type data`` definitions to define constructors
+at the type level, but programmers using these definitions may also want
+to use type-level literals.
+
+Proposed Change Specification
+-----------------------------
+
+The extension ``TypeLevelLiterals`` would enable type-level literals,
+as described in the *Type-Level Literals* section of the *User's Guide*.
+The new extension would be implied by the ``DataKinds`` extension, yielding
+backwards compatibility.
+
+Examples
+--------
+
+A sized vector type without promoting the constructors to the type level::
+
+  {-# LANGUAGE TypeLevelLiterals, GADTs #-}
+  module Vector where
+
+  import Data.Kind (Type)
+  import GHC.TypeLits
+  import Numeric.Natural
+
+  data Vector :: Natural -> Type -> Type where
+    Nil :: Vector 0 a
+    Cons :: a -> Vector n a -> Vector (n+1) a
+
+Effect and Interactions
+-----------------------
+Modules containing::
+
+  {-# LANGUAGE TypeLevelLiterals #-}
+
+will be able to use integer, string and character literals in types,
+without having all their data constructors promoted into the type
+constructor namespace (unless requested).
+
+Modules containing::
+
+  {-# LANGUAGE DataKinds #-}
+
+will behave as before, because ``DataKinds`` will imply ``TypeLevelLiterals``.
+
+Costs and Drawbacks
+-------------------
+The cost is an extra language extension name.
+It is trivial to implement.
+
+Alternatives
+------------
+The existing workaround of using ``DataKinds`` may add unwanted names to
```

Comment:
This phrasing is odd to me. `DataKinds` does not add anything to the type constructor namespace. If you declare

```haskell
data T = T
```

then you have `T` the type constructor and `T` the data constructor, each in its own namespace. That is why the `'T` syntax exists — to select the data namespace when writing code at the type level.

I think you wanted to say something along the lines of "The existing extension `DataKinds` allows the use of data constructors at the type level, creating name conflicts with type constructors".

Reply:
Name conflicts are part of it. With`DataKinds`, a declaration
```haskell
data T = C
```
allows `C` to be used as a data constructor in terms and as a type constructor in types (if no other type constructor `C` is defined). Without `DataKinds`, `C` is only a data constructor, and using it in a type is reported as an error.

Reply:
> allows `C` to be used as a data constructor in terms and as a type constructor in types 

That's where we disagree on terminology. I say `C` is a data constructor, period. It may be used as a data constructor in terms (naturally) and as a data constructor in types (with `DataKinds`).

Reply:
I was following the _User's Guide_, which says `DataKinds` promotes "(value) constructors to be type constructors".

Reply:
OK, now I see where you're coming from. While I believe the User's Guide is incorrect in this case and conflates type constructors with type-level usage of data constructors, I realize that this proposal isn't the best place to debate it, so consider my objection withdrawn.